Migrate to flexmark-java

ikappaki opened the pull request:
Hi,

can you please review draft patch for migrating PegDown to flexmark-java solves #158 as well as #197. I've also upgraded to latest Clojure(Script) versions while at it.

The codex/example project generates markdown docs as expected, so functionality wise the patch appears to be complete. I still need to do a bit of tidying up (around the three reification functions, there is some performance cost around it), I thought to send it out for review first to get some feedback I'm on the right track.

I went mostly out for compatibility with the existing PegDown code. The only think I don't understand and haven't ported over is the link `encode-title`. Why do we needed to encode the link title in the first place as opposed to everything else that we don't? The browser appears to display it correctly in both cases.

Tested the codox/example project to work with all Java 8, 11 and 17.

Thanks.

weavejester commented:
Thanks for the patch. I believe `encode-title` was a workaround for a problem with Pegdown, so it probably doesn't apply to Flexmark. It looks like this PR is going to take a while to review, but it's a necessary upgrade, so I'll see if I can find a few hours within the next week.

ikappaki commented:
(I've tidied up the code a little:  project picks up only the required subset of flexmark libs while nested the reifies)

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
+         (affectsGlobalScope [this] false)
+         (^LinkResolver apply [this ^LinkResolverBasicContext BasicContetcontext]
+          (reify LinkResolver
+            (resolveLink [this node context link]
```
Unused arguments names should be changed to `_`.

ikappaki replied:
Done.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
+                :else
+                link)))))))))
+
+(defn- flexmark-options-make  [project & [ns]]
```
A better name would be `make-flexmark-options`.

ikappaki replied:
Done.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
+      (.toImmutable)))
+
+(defn- markdown-to-html [doc project & [ns]]
+  (let [options (flexmark-options-make project ns)
```
The let clauses should be aligned for clarity. e.g.

```clojure
  (let [options  (flexmark-options-make project ns)
        parser   (.build (Parser/builder options))
        renderer (.build (HtmlRenderer/builder options))
        doc      (.parse parser doc)]
    (.render renderer doc))
```

I think it also might be clearer as:

```clojure
  (let [options  (flexmark-options-make project ns)
        parser   (.build (Parser/builder options))
        renderer (.build (HtmlRenderer/builder options))]
    (->> doc (.parse parser) (.render renderer)))
```

ikappaki replied:
Used the clearer version.

weavejester commented on example/src/clojure/codox/markdown.clj:
```diff
@@ -31,9 +31,9 @@
   ~~~
 
-  [Inline links](http://example.com) and [reference links][1].
+  [Inline links](http://example.com 'Inline') and [reference links][1].
```
This change doesn't seem related to flexmark as far as I can see.

ikappaki replied:
I've updated the links while investigating the old link topic fixes, and I thought it was a good idea to have them in the examples. I've reverted the change.

weavejester reviewed (changes requested):
Thanks for the cleanup. That made the code a lot better. I still have a list of suggestions, and I've yet to actually test it at all, but so far it looks promising.

ikappaki commented on codox/project.clj:
```diff
@@ -4,10 +4,12 @@
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.7.0"]
+  :dependencies [[org.clojure/clojure "1.10.3"]
```
It's not, I though it was a good idea to begin with, but is better to keep them separate for clarity. I've reverted the update.

ikappaki commented on codox/src/codox/writer/html.clj:
```diff
@@ -2,11 +2,14 @@
-           [org.pegdown.ast ExpLinkNode RefLinkNode WikiLinkNode])
+           [com.vladsch.flexmark.ast Link LinkRef]
+           [com.vladsch.flexmark.ext.wikilink WikiLink WikiLinkExtension]
+           [com.vladsch.flexmark.html HtmlRenderer HtmlRenderer$HtmlRendererExtension
```
Done.

ikappaki commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
-         (-> (LinkRenderer$Rendering. (fix-markdown-url url) text)
-             (encode-title title))
-         (proxy-super render node url title text))))))
+;;; Adapted from https://github.com/vsch/flexmark-java/blob/8a881b73109a287b5f202e2e1fc9f9c497d5eecf/flexmark-java-samples/src/com/vladsch/flexmark/java/samples/PegdownCustomLinkResolverOptions.java.
```
Removed. I will consider adding it to the commit message while squashing at the end.

ikappaki commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
-             (encode-title title))
-         (proxy-super render node url title text))))))
+;;; Adapted from https://github.com/vsch/flexmark-java/blob/8a881b73109a287b5f202e2e1fc9f9c497d5eecf/flexmark-java-samples/src/com/vladsch/flexmark/java/samples/PegdownCustomLinkResolverOptions.java.
+(defn- CustomExtension-make [project & [ns]]
```
Switched to multi-arity functions.

ikappaki commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +82,66 @@
+         (^LinkResolver apply [this ^LinkResolverBasicContext BasicContetcontext]
+          (reify LinkResolver
+            (resolveLink [this node context link]
+              ;; correct internal link references.
```
Done.

ikappaki commented:
Hi and thank you very much for the review, I've updated the codebase as per review comments. Let me know if you want me to squash and force  the commits instead of doing incremental code updates.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +84,75 @@
+               (correct-internal-links node link project ns))))))))))
+
+(defn- make-flexmark-options
+  ([project]
+   (make-flexmark-options project nil))
```
Since this is a private function, and never called without the `ns` argument that I can see, we can probably just have one arity.

ikappaki replied:
good catch, converted to single arity.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +84,75 @@
+    link))
+
+(defn- make-renderer-extension
+  ([project]
+   (make-renderer-extension project nil))
```
Since this is a private function, and never called without the ns argument, we can probably just have one arity.

ikappaki replied:
converted to single arity.

weavejester commented on codox/src/codox/writer/html.clj:
```diff
@@ -100,34 +84,75 @@
-         (-> (LinkRenderer$Rendering. (fix-markdown-url url) text)
-             (encode-title title))
-         (proxy-super render node url title text))))))
+(defn- correct-internal-links [node link project ns]
```
What about:

```clojure
(defn- update-link-url [^ResolvedLink link f]
  (-> link
      (.withStatus LinkStatus/VALID)
      (.withUrl (f (.getUrl link)))))

(defn- correct-internal-links [node link project ns]
  (condp instance? node
    WikiLink (update-link-url link #(find-wiki-link project ns %))
    LinkRef  (update-link-url link fix-markdown)
    Link     (update-link-url link fix-markdown)
    link))

```

ikappaki replied:
Prettier, code updated.

ikappaki commented:
Have just incorporated latest review comments, thanks.

weavejester commented:
Sorry for the delay in reviewing this. I've tested it out on the example project, and it's all looks okay as far as I've been able to determine. The code is fine now.

Could you squash your commits down and give them a commit message in line with [Chris Beam's seven rules](https://chris.beams.io/posts/git-commit/#seven-rules). So perhaps something like:

```
Migrate markdown rendering to Flexmark

Pegdown has been deprecated.
```

ikappaki commented:
Hi, thanks for your time going through the iterative reviews. 

I have squashed the commits and added a commit message as suggested (also included the reference to the example java code I've used to migrate to flexmark -- in case it becomes useful for future reference or troubleshooting).

weavejester commented:
Thanks! I forgot one thing: can you add `Fixes #158 and #197.` to the commit message?

ikappaki commented:
> 
> 
> Thanks! I forgot one thing: can you add `Fixes #158 and #197.` to the commit message?

Done.

weavejester commented:
Thanks, but can you remove the `weavejester` part in front of the two issue numbers? It's unnecessary.

ikappaki commented:
> 
> 
> Thanks, but can you remove the `weavejester` part in front of the two issue numbers? It's unnecessary.

I believe this is an artefact of the github web viewer? the actual commit message (I've just double checked on my machine)  starts with the hash symbol followed by the issue number.

(Or is it that github rewritten the log message after I pushed it? Not sure how to rewrite it if so. )

weavejester commented:
Ah, so it is! When I take a look at the commit via the commandline, it doesn't show those extra elements. That should all be fine, then, so let me merge it in now.
